### PR TITLE
Document behaviour_info/1

### LINKED ISSUE #13733: Make `behaviour_info` more accessible in documentation
Right now, if you want to get information about the `behaviour_info/1` function that gets generated in behaviours, you have to do one of the following:

  * Go to the docs for the `Behaviour` module. This is bad as that module is *deprecated*.
  * Google `behaviour_info elixir`, which returns sucky results (screenshot below)
  * Google `behaviour_info erlang`, where for me the **third result** is the [*Modules* Erlang docs](https://www.erlang.org/docs/17/reference_manual/modules), where then I have to search `behaviour_info` in the page.

This is all not ideal IMO, can we make it better? I’m a bit out of ideas. We could implement a mock `Kernel.behaviour_info/1` function just to document stuff, but I’m not sure that's a good solution. Thoughts? 

CleanShot 2024-07-23 at 16 02 31@2x

### PR DESCRIPTION
**Pull request by whatyouhide**
Closes #13733.

### CODE REVIEW

File: lib/elixir/lib/module.ex [hunk 1]
```diff
@@ -590,6 +591,40 @@ defmodule Module do
       `@compile {:no_warn_undefined, {Mod, fun, arity}}` - does not warn if
       the given module or the given `Mod.fun/arity` are not defined
 
+  ## Generated Functions
+
+  Sometimes the compiler will generated public functions within modules. These
+  are documented below.
```
Comment (josevalim) [inline lib/elixir/lib/module.ex:597]: Should we document them `module_info` generated by Erlang `__info__` generated by Elixir?

File: lib/elixir/lib/module.ex [hunk 2]
```diff
@@ -590,6 +591,40 @@ defmodule Module do
       `@compile {:no_warn_undefined, {Mod, fun, arity}}` - does not warn if
       the given module or the given `Mod.fun/arity` are not defined
 
+  ## Generated Functions
+
+  Sometimes the compiler will generated public functions within modules. These
```
Comment (davydog187) [inline lib/elixir/lib/module.ex:596]:
```suggestion
  Sometimes the compiler will generate public functions within modules. These
```

Comment (lukaszsamson) [reply lib/elixir/lib/module.ex:596]: Does't sometimes sound like things happen at random?

Comment (wojtekmach) [reply lib/elixir/lib/module.ex:596]: Yeah, perhaps:

> In certain situations the compiler will (...)

File: lib/elixir/lib/module.ex (hunk 1 above)
Comment (whatyouhide) [reply lib/elixir/lib/module.ex:597]: Yep, added!

File: lib/elixir/lib/module.ex (hunk 2 above)
Comment (whatyouhide) [reply lib/elixir/lib/module.ex]: Nice catch, fixed!

File: lib/elixir/lib/module.ex [hunk 3]
```diff
@@ -590,6 +591,78 @@ defmodule Module do
...
 
+  ## Generated Functions
+
+  Sometimes the compiler will generate public functions within modules. These
+  are documented below.
+
+  ### `module_info/1`
+
+  This function is generated for all modules and returns a keyword list with
+  information about the module. The signature for this function, expressed as
```
Comment (josevalim) [inline lib/elixir/lib/module.ex:602]: It does not return a keyword list :)

File: lib/elixir/lib/module.ex [hunk 4]
```diff
@@ -590,6 +591,78 @@ defmodule Module do
...
+
+  ### `module_info/0`
+
+  This function is generated for all modules. It returns all the attributes
+  returned by `module_info/1`, but as a single keyword list. See also the
+  [Erlang documentation](https://www.erlang.org/doc/system/modules.html#module_info-0).
+
+  ### `__info__/1`
+
+  This function is generated for all modules. It's similar to `module_info/1` but
```
Comment (josevalim) [inline lib/elixir/lib/module.ex:633]:
```suggestion
  This function is generated for all modules defined by Elixir. It's similar to `module_info/1` but
```

File: lib/elixir/lib/module.ex [hunk 5]
```diff
@@ -590,6 +591,78 @@ defmodule Module do
...
+  includes some additional Elixir-specific information, such as struct and macro
+  information. For documentation, see `c:Module.__info__/1`.
+
+  ### `behaviour_info/1`
```
Comment (josevalim) [inline lib/elixir/lib/module.ex:637]: WDYT about alphabetical order with `__info__` at the bottom? :D

**Pull request merged**